### ISSUE TITLE
Clicking ToC link within an already open document doesn't work

### ISSUE THREAD

**Issue by becka11y** [readium/readium-js-viewer#549]
When in Document or Continuous mode, clicking on a ToC link for a document that is already loaded will display the "busy spinner" indefinitely. The view is not advanced to the new location.

#### Related issue(s) 
This was discovered while I was researching #543. I have closed that issue since it has nothing to do with google generated docs. 

#### Expected Behavior
When I click on any ToC link, I expect the associated page/section to be displayed by Readium.

#### Observed behavior
Readium does not navigate to the selected page/section from the ToC if the document with that page/section is already loaded in memory. 

#### Steps to reproduce
It is easiest to reproduce this with layout scroll mode set to document but it can also be done in continuous mode.
1. From the library view open Settings and select the layout tab and select document as the scroll mode (display mode settings doesn't matter).
2.  Open the Accessible EPUB3 book. 
3,  Open the Table of Contents
4. Click on "2. Building a Better EPUB: Fundamental Accessibility" to load chapter 2
Now click on any of the other links in chapter 2 - that page is never loaded, the display doesn't update and the spinner keeps spinning.
Basically any ToC link that is relative (contains a #anchorRef) to the already loaded page will not work.  Only a completely new url will load properly when selected from the ToC. 

#### Test file(s)
Happens on all books that I tested.  

### Product

latest official version available from the Chrome Web Store - 2.23.0
Version of the Chrome browser - Version 51.0.2704.103 (64-bit)
OS and version - OS X 10.11.5

Readium cloud reader app
2.24.0-alpha
Wed, 15 Jun 2016 15:16:04 GMT
readium-js-viewer@39939d8
readium-js@b32f9eb
readium-shared-js@91324db
readium-cfi-js@a9b07a7
Browser and version - Safari Version 9.1.1 (11601.6.17)
OS and version - OS X 10.11.5

**Comment by rkwright** [comment]
@becka11y A couple of clarifications.  By "Document" you mean scroll-document mode.  I was unable to trigger it in "paginated" mode but was able to reproduce in "scroll-doc" mode as well as less so in continuous mode.  But I do note that the reader isn't "hung" per se, it just has the spinner running and that can be cleared by by navigating somewhere else.  Still, a nasty bug.

**Comment by becka11y** [comment]
I'm not quite sure what you mean by pagination mode? I could reproduce this if I picked document or continuous scroll mode in settings layout panel. The display mode setting had no effect. I also found it harder to reproduce in continuous mode since pages get auto loaded and I'm not sure how or if they get switched out if the whole ePub can't get loaded. 

Right, the reader isn't hung. You can also remove the spinner by closing the ToC.

**Comment by rkwright** [comment]
@becka11y I meant only that I could only get it to happen in one of the two scroll modes.  I couldn't seem to reproduce it when in paginated (non-scroll) mode.

**Comment by becka11y** [comment]
I started to debug this.  The problem is in scroll_view.js with a call to sfi.getVerticalOffsetForElement in the openPageViewElement(pageView, pageRequest) function.  

The  getVerticalOffsetForElement function was removed from cfi_navigation_logic.js in commit:      a110d5fada84bea8213fb8702acbdbcc314b4224 

That commit was fixing a similar problem. See: https://github.com/readium/readium-shared-js/issues/253

**Comment by rkwright** [comment]
This may be the same bug as well:

http://idpf.org/forum/topic-3129

**Comment by becka11y** [comment]
Adding back the two deleted functions seems to fix the problems. I just have to refresh my aging memory on how to deal with submodules in order to commit a branch for review and testing. 

    this.getVerticalOffsetForElement = function ($element) {		
      return this.getVerticalOffsetForPointOnElement($element, 0, 0);		
    };
    
    this.getVerticalOffsetForPointOnElement = function ($element, x, y) {		
      var elementRect = Helpers.Rect.fromElement($element);		
      return Math.ceil(elementRect.top + y * elementRect.height / 100);		
    };

**Comment by danielweck** [comment]
Fixed by https://github.com/readium/readium-shared-js/pull/309
